File: client-scalaz/src/main/scala/org/openapitools/client/api/Route.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import Route._

case class Route (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
name: Option[String],
externalId: Option[String],
hub: Option[ServiceHub],
program: Option[Program],
zone: Option[Region],
vehicle: Option[Vehicle],
driver: Option[Account],
concierge: Option[Account],
scheduled: Option[OffsetDateTime],
approved: Option[OffsetDateTime],
started: Option[OffsetDateTime],
completed: Option[OffsetDateTime],
totalDistance: Option[Float],
totalTime: Option[Long],
estimatedDistance: Option[Double],
estimatedTime: Option[Long],
startStop: Option[Stop],
endStop: Option[Stop],
stops: Option[List[Stop]],
shipments: Option[List[Shipment]],
allShipmentsConfirmed: Option[Boolean],
shipmentOrderCount: Option[Integer],
polyline: Option[String],
notes: Option[List[Note]],
noteCount: Option[Long],
glympseToken: Option[String],
enableOptimization: Option[Boolean],
optimizedDate: Option[OffsetDateTime],
optimizationStarted: Option[OffsetDateTime],
optimizationCompleted: Option[OffsetDateTime],
featured: Option[Boolean],
image: Option[Asset],
parent: Option[Route],
contentName: Option[String])

object Route {
  import DateTimeCodecs._

  implicit val RouteCodecJson: CodecJson[Route] = CodecJson.derive[Route]
  implicit val RouteDecoder: EntityDecoder[Route] = jsonOf[Route]
  implicit val RouteEncoder: EntityEncoder[Route] = jsonEncoderOf[Route]
}

File: client-clojure/src/sirqul_io_t_platform/specs/route.clj
(ns sirqul-io-t-platform.specs.route
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.service-hub :refer :all]
            [sirqul-io-t-platform.specs.program :refer :all]
            [sirqul-io-t-platform.specs.region :refer :all]
            [sirqul-io-t-platform.specs.vehicle :refer :all]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.stop :refer :all]
            [sirqul-io-t-platform.specs.stop :refer :all]
            [sirqul-io-t-platform.specs.stop :refer :all]
            [sirqul-io-t-platform.specs.shipment :refer :all]
            [sirqul-io-t-platform.specs.note :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.route :refer :all]
            )
  (:import (java.io File)))


(def route-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :name) string?
   (ds/opt :externalId) string?
   (ds/opt :hub) service-hub-spec
   (ds/opt :program) program-spec
   (ds/opt :zone) region-spec
   (ds/opt :vehicle) vehicle-spec
   (ds/opt :driver) account-spec
   (ds/opt :concierge) account-spec
   (ds/opt :scheduled) inst?
   (ds/opt :approved) inst?
   (ds/opt :started) inst?
   (ds/opt :completed) inst?
   (ds/opt :totalDistance) float?
   (ds/opt :totalTime) int?
   (ds/opt :estimatedDistance) float?
   (ds/opt :estimatedTime) int?
   (ds/opt :startStop) stop-spec
   (ds/opt :endStop) stop-spec
   (ds/opt :stops) (s/coll-of stop-spec)
   (ds/opt :shipments) (s/coll-of shipment-spec)
   (ds/opt :allShipmentsConfirmed) boolean?
   (ds/opt :shipmentOrderCount) int?
   (ds/opt :polyline) string?
   (ds/opt :notes) (s/coll-of note-spec)
   (ds/opt :noteCount) int?
   (ds/opt :glympseToken) string?
   (ds/opt :enableOptimization) boolean?
   (ds/opt :optimizedDate) inst?
   (ds/opt :optimizationStarted) inst?
   (ds/opt :optimizationCompleted) inst?
   (ds/opt :featured) boolean?
   (ds/opt :image) asset-spec
   (ds/opt :parent) route-spec
   (ds/opt :contentName) string?
   })

(def route-spec
  (ds/spec
    {:name ::route
     :spec route-data}))
